File: scan_git/src/test/java/common/JGitTest.java
package common;

import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.api.errors.GitAPIException;
import org.eclipse.jgit.diff.DiffEntry;
import org.eclipse.jgit.diff.DiffFormatter;
import org.eclipse.jgit.diff.RawTextComparator;
import org.eclipse.jgit.lib.FileMode;
import org.eclipse.jgit.lib.ObjectReader;
import org.eclipse.jgit.lib.Repository;
import org.eclipse.jgit.revwalk.RevCommit;
import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
import org.eclipse.jgit.treewalk.CanonicalTreeParser;
import org.eclipse.jgit.treewalk.TreeWalk;
import org.eclipse.jgit.util.io.DisabledOutputStream;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.util.List;

class JGitTest {


    private Repository repository;

    @BeforeEach
    void setUp() throws IOException {
        repository = new FileRepositoryBuilder()
                .findGitDir(new File("/Users/ygdong/Downloads/gittest"))
                .build();
    }




    /*最终采用这种方式*/
    @Test
    void diffFormat() throws GitAPIException, IOException {
        try (Git git = new Git(repository)) {
            DiffFormatter formatter = new DiffFormatter(DisabledOutputStream.INSTANCE);
            formatter.setRepository(repository);
            formatter.setDiffComparator(RawTextComparator.DEFAULT);
            formatter.setDetectRenames(true);
            int count = 0;
            for (RevCommit commit : git.log().call()) {
                count++;
                System.out.println("commit:" + commit.getName() + ", message:" + commit.getFullMessage());
                RevCommit parent = commit.getParentCount() != 0 ? commit.getParent(0) : null;
                List<DiffEntry> diffEntries = formatter.scan(parent == null ? null : parent.getTree(), commit.getTree());
                diffEntries.forEach(e ->
                        System.out.println("changeType:=" + e.getChangeType().name() + ",newID=" + e.getNewId() + ",newPath=" + e.getNewPath()));
            }
            System.out.println("count = " + count);
        }
    }


    /*这是第二种方式，也可以*/
    @Test
    void canonicalTreeParserTest() throws GitAPIException, IOException {
        try (Git git = new Git(repository)) {
            try (ObjectReader reader = repository.newObjectReader()) {
                CanonicalTreeParser oldTreeIter = new CanonicalTreeParser();
                CanonicalTreeParser newTreeIter = new CanonicalTreeParser();
                for (RevCommit commit : git.log().call()) {
                    System.out.println("commit:" + commit.getName() + ", message:" + commit.getFullMessage());
                    RevCommit parent = commit.getParentCount() != 0 ? commit.getParent(0) : null;
                    if (parent != null)
                        oldTreeIter.reset(reader, parent.getTree());
                    newTreeIter.reset(reader, commit.getTree());
                    List<DiffEntry> diffEntries = git.diff().setNewTree(newTreeIter).setOldTree(oldTreeIter).call();
                    diffEntries.forEach(e ->
                            System.out.println("changeType:=" + e.getChangeType().name() + ",newID=" + e.getNewId() + ",newPath=" + e.getNewPath()));
                }
            }
        }
    }

    /*not work，因为会包含一个所有的文件，不仅仅是变更的文件*/
    @Test
    void testTreeWalk() throws GitAPIException, IOException {
        try (Git git = new Git(repository); TreeWalk treeWalk = new TreeWalk(repository)) {
            Iterable<RevCommit> revCommits = git.log().call();
            for (RevCommit commit : revCommits) {
                treeWalk.reset(commit.getTree());
                System.out.println("commit:" + commit.getName() + ", message:" + commit.getFullMessage());
                while (treeWalk.next()) {
                    FileMode fileMode = treeWalk.getFileMode();
                    System.out.println("fileMode = " + fileMode.toString());
                    String pathString = treeWalk.getPathString();
                    System.out.println("pathString = " + pathString);

                }
            }
        }
    }
}

File: scan_git/src/test/kotlin/JGitTest.kt
import org.eclipse.jgit.api.Git
import org.eclipse.jgit.api.errors.GitAPIException
import org.eclipse.jgit.diff.DiffEntry
import org.eclipse.jgit.diff.DiffFormatter
import org.eclipse.jgit.diff.RawTextComparator
import org.eclipse.jgit.lib.Repository
import org.eclipse.jgit.storage.file.FileRepositoryBuilder
import org.eclipse.jgit.treewalk.CanonicalTreeParser
import org.eclipse.jgit.treewalk.TreeWalk
import org.eclipse.jgit.util.io.DisabledOutputStream
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.File
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.function.Consumer

class JGitTest {

    private var repository: Repository? = null

    @BeforeEach
    @Throws(IOException::class)
    fun setUp() {
        repository = FileRepositoryBuilder()
                .findGitDir(File("test_data"))
                .build()
    }


    /*最终采用这种方式*/
    @Test
    @Throws(GitAPIException::class, IOException::class)
    fun diffFormat() {
        Git(repository).use { git ->
            val formatter = DiffFormatter(DisabledOutputStream.INSTANCE)
            formatter.setRepository(repository)
            formatter.setDiffComparator(RawTextComparator.DEFAULT)
            formatter.isDetectRenames = true
            var count = 0
            for (commit in git.log().call()) {
                count++
                println("commit:" + commit.name + ", message:" + commit.fullMessage)
                val parent = if (commit.parentCount != 0) commit.getParent(0) else null
                val diffEntries = formatter.scan(parent?.tree, commit.tree)
                diffEntries.forEach(Consumer { e: DiffEntry -> println("changeType:=" + e.changeType.name + ",newID=" + e.newId + ",newPath=" + e.newPath) })
            }
            println("count = $count")
        }
    }


    /*这是第二种方式，也可以*/
    @Test
    @Throws(GitAPIException::class, IOException::class)
    fun canonicalTreeParserTest() {
        Git(repository).use { git ->
            repository!!.newObjectReader().use { reader ->
                val oldTreeIter = CanonicalTreeParser()
                val newTreeIter = CanonicalTreeParser()
                for (commit in git.log().call()) {
                    println("commit:" + commit.name + ", message:" + commit.fullMessage)
                    val parent = if (commit.parentCount != 0) commit.getParent(0) else null
                    if (parent != null) oldTreeIter.reset(reader, parent.tree)
                    newTreeIter.reset(reader, commit.tree)
                    val diffEntries = git.diff().setNewTree(newTreeIter).setOldTree(oldTreeIter).call()
                    diffEntries.forEach(Consumer { e: DiffEntry -> println("changeType:=" + e.changeType.name + ",newID=" + e.newId + ",newPath=" + e.newPath) })
                }
            }
        }
    }

    /*not work，因为会包含一个所有的文件，不仅仅是变更的文件*/
    @Test
    @Throws(GitAPIException::class, IOException::class)
    fun testTreeWalk() {
        Git(repository).use { git ->
            TreeWalk(repository).use { treeWalk ->
                val revCommits = git.log().call()
                for (commit in revCommits) {
                    treeWalk.reset(commit.tree)
                    println("commit:" + commit.name + ", message:" + commit.fullMessage)
                    while (treeWalk.next()) {
                        val fileMode = treeWalk.fileMode
                        println("fileMode = $fileMode")
                        val pathString = treeWalk.pathString
                        println("pathString = $pathString")
                    }
                }
            }
        }
    }


    /** @see https://stackoverflow.com/questions/45793800/jgit-read-the-content-of-a-file-at-a-commit-in-a-branch*/
    @Test
    internal fun getContent_for_path_in_aCommit() {
        Git(repository).use { git ->
            git.log().call().forEach { revCommit ->
                println(revCommit.fullMessage + "++++++++++++++")
                TreeWalk.forPath(repository, "b", revCommit.tree).use { treeWalk ->
                    if (treeWalk != null) {
                        val objectId = treeWalk.getObjectId(0)
                        repository!!.newObjectReader().use { objectReader ->
                            val bytes = objectReader.open(objectId).bytes
                            println(String(bytes, StandardCharsets.UTF_8))
                        }
                    }
                }
            }
        }
    }


}